libyul/optimiser/Semantics.cpp

```diff
@@ -93,6 +94,35 @@ void MSizeFinder::operator()(FunctionCall const& _functionCall)
 			m_msizeFound = true;
 }
 
+
+map<YulString, SideEffects> SideEffectsPropagator::sideEffects(
+	Dialect const& _dialect,
+	map<YulString, std::set<YulString>> const& _directCallGraph
+)
+{
+	map<YulString, SideEffects> ret;
+	for (auto const& call: _directCallGraph)
```

Comment:
```suggestion
	for (auto const& [funName,dummy]: _directCallGraph)
```
:-)?

Reply:
Instead of ``dummy`` we could probably even use ``directCalls`` and use that to initialize ``BreadthFirstSearch<YulString>{{directCalls}}`` and change the visit to
```
{
	if (sideEffects == SideEffects::worst())
		return;
	if (BuiltinFunction const* f = _dialect.builtin(_function))
		sideEffects += f->sideEffects;
	else for (YulString callee: _directCallGraph.at(_function))
			_addChild(callee);
}
```

Reply:
It seems my compiler is not yet fully c++17-compliant...

Reply:
I don't understand what you mean by `directCalls`.

Reply:
Without the C++17 change that would be ``call.second`` - the idea was not to start the traversal at the function itself (i.e. at ``funName``), but to start directly with the list of its callee's (i.e. ``call.second``, resp. ``directCalls``) and change the visitor accordingly.

In total my suggestion would be
```
	map<YulString, SideEffects> ret;
	for (auto const& [funName, directCalls]: _directCallGraph)
	{
		SideEffects sideEffects;
		BreadthFirstSearch<YulString>{directCalls}.run(
			[&](YulString _function, auto&& _addChild) {
				if (sideEffects == SideEffects::worst())
					return;
				if (BuiltinFunction const* f = _dialect.builtin(_function))
					sideEffects += f->sideEffects;
				else for (YulString callee: _directCallGraph.at(_function))
					_addChild(callee);
			}
		);
		ret[funName] = sideEffects;
	}
```

Reply:
Oh I see!

Reply:
I.e. in the outer loop we not only get all functions as keys, but already have all of their direct calls in the image of the map, so we might as well start traversing at those directly.